### ISSUE TITLE
Switch HTTP server away from thin.

### ISSUE THREAD

**Issue by andrewvc** [andrewvc/dripdrop#4]
Thin's a great server, but it doesn't work with jruby. Boo. perhaps evma-httpserver?

**Comment by wishdev** [comment]
Have you considered using Mongrel2 (http://mongrel2.org/home) as the preferred option for HTTP access. It's fully 0MQ capable along with a whole host of adapters to any language you can think of.

It might be easier to have good instructions on how to marry Mongrel2 in then try to re-invent that wheel

**Comment by andrewvc** [comment]
Hmm, I think you're right. I'm going to leave HTTP in as implemented now, because it does work, but mongrel2 definitely be a better 'production' solution.

I'll have to take another look at the m2r adapter. It's not evented (uses ffi-rzmq directly, no zmqmachine) so far as I know. Maybe I can just rip out/or include the actual protocol handling parts as a dripdrop handler since it's just push/pull and pub/sub.

Also, I'll have to write specs for push/pull in zmqmachine so they'll get accepted upstream, I can't expect people to build my custom branch forever. So it's a little bit of a road.

**Comment by hedgehog** [comment]
I haven't got my hands too dirty with the code just yet... but is there any reason you couldn't use a rack handler?

https://github.com/darkhelmet/rack-mongrel2

**Comment by andrewvc** [comment]
That would be ideal, do you have any knowledge as to whether https://github.com/rkh/async-rack supports the rack-mongrel2 handler? I haven't used either.

**Comment by hedgehog** [comment]
I bumped into aysnc_rack (and rack-test) while trying to get to grips with async-sinatra... about a year ago, or maybe the night mare just seems further back.

I thought aysnc-sinatra was terrible - to be fair some openly admitted it was a cludge and had to be re-done at some point.  The thing that was bad about it was the throwing and catching to handle events, and this is what async-rack does too.  It seems it is a matter of religion/taste....

If you do choose async via a throw-catch route, can it be made optional?  i.e. something no user ever has to do?  And have it implemented as an included module so that alternative approaches can be dropped in.

Final whinge point. Rack-test. If you do go down the rack path, please can Rack-test never enter into DripDrop.

**Comment by andrewvc** [comment]
Well, you seem pretty interested in a non-reactor version of drip drop. I think anything non-reactor should not go in the Node class, but perhaps be part of a new Threaded class.

**Comment by hedgehog** [comment]
> Well, you seem pretty interested in a non-reactor version of drip drop

I could have my understanding out of sync with the formal comp sci definitions.
I'm not opposed to a reactor pattern, it just got very tricky the last time I used it. Specifically,  something like  

```
[msg_id] = proc_lambda
```

Then grab-and-run as messages come in, there was I think some sequencing enforced so we actually used pop-and-run, and from distant memory a hash was tried for some things.  At no time did we throw or catch anything. I understood this to be the reactor pattern - apologies if that is not the case.

I then switched to using, what I think is the observer 'pattern' (or Aspect oriented programming/pattern), where we didn't have to worry about building/managing hashes or arrays of proc/lambdas.  BUT unit spec's and (before cucumber) integrations specs were essential.

Hope that clarifies what I meant?
